Simple theme: implement RTL

dalf opened the pull request:
## What does this PR do?

This PR implements the RTL text direction in the simple theme.

Modern browsers do most of the work, but margin / padding / border and some other properties have to be manage manually.

The usual method seems to be:
```css
#q {
  border-right: none;
}

#q [dir="rtl] {
  border-right: inherit;
  border-left: none;
}
```

IMO, this is error prone and require some copy / paste 

There is the [`margin-inline-start`](https://developer.mozilla.org/en-US/docs/Web/CSS/margin-inline-start ) property, but it seems to be supported only recently (except for FF), and it doesn't solve all cases (for example `float: right`).

So this PR uses some LESS mixins:

In LTR version:
```less
.ltr-left(@offset) {
  left: @offset;
}
```

In this RTL version:
```less
.ltr-left(@offset) {
  right: @offset;
}
```

and so on for the different CSS properties.

## Why is this change important?

Add support for RTL languages in the Simple theme.

![image](https://user-images.githubusercontent.com/1594191/149620444-ca2700e7-1f1f-4ecd-a7d3-88b77a28ff0f.png)

## How to test this PR locally?

* check everything in a LTR language
* check everything in a RTL language

## Author's checklist

Issue list:
* [x] URL in the result
* [x] Vim hotkeys help
* [x] LTR text in the result are displayed correctly.

For the last one: I dont' know if there is a way to display LTR text aligned on the left, and RTL text aligned on the right only using CSS.

Currently, the only way I see is to [detect the text direction in Python](https://docs.python.org/3.8/library/unicodedata.html#unicodedata.bidirectional), and then add the `dir` HTML attribute.

From that, it is possible to add some CSS to display the text either aligned on the right, either aligned on the left.

ping @GreenLunar 

## Related issues

Close https://github.com/searxng/searxng/issues/747

dalf committed:
[enh] simple theme: RTL support

* mirror all inline SVGs so that direction SVGs display correctly on RTL
* set the bold list element in info box to RTL so the colon gets displayed on the right side
* set correct .ltr function for the left border on the search button in #q
* move text to the right in autocomplete
* move search form in lign with result article on RTL
* add the correct padding for img thumbnails in categories like music on RTL
* apply RTL to result table for map results
* align text in tables part of /preferences on RTL
* move burger menu on index page to the left on RTL
* fix positioning of drop down arrow on select boxes on RTL
* align result URL on the right (written LTR)
* align vim hotkeys help on the left since it is not translated
* image detail:
  * labels (author, format, URL, etc...) are written on the right,
    values are on the left.
  * URL are written LTR and overflow on the right

return42 commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -117,7 +117,7 @@ module.exports = function (grunt) {
         },
         files: {
-          "css/searxng.min.css": "src/less/style.less",
+          "css/searxng.min.css": "src/less/style-ltr.less",
```
I miss the `src/less/style-ltr.less` file .. or do I misunderstand this?

```
Running "less:production" (less) task
>> Destination css/searxng.min.css not written because no source files were found.
>> 1 stylesheet created.
>> 1 sourcemap created.
```
I wondering that the build `make themes.simple` ends without an error .. :confused:

dalf replied:
Face palm, I forgot to add this file. Fixed.

return42 commented on searx/static/themes/simple/.stylelintrc.json:
```diff
@@ -1,3 +1,7 @@
 {
-  "extends": "stylelint-config-standard"
+  "extends": "stylelint-config-standard",
+  "rules": {
+    "declaration-empty-line-before": null,
+    "no-invalid-position-at-import-rule": null
+  }
```
IMO this should be done in a separated commit, with a commit message why we want to activate:

- https://stylelint.io/user-guide/rules/list/declaration-empty-line-before/
- https://stylelint.io/user-guide/rules/list/no-invalid-position-at-import-rule/

dalf replied:
With `declaration-empty-line-before` enabled:
```less
#results.image-detail-open article.result-images[data-vim-selected] .detail {
  display: flex;
  flex-direction: column;
  position: fixed;
  .ltr-left(60rem);
  .ltr-right(0);
  top: @search-height;
```

has to be written
```less
#results.image-detail-open article.result-images[data-vim-selected] .detail {
  display: flex;
  flex-direction: column;
  position: fixed;

  .ltr-left(60rem);

  .ltr-right(0);
  top: @search-height;
```

---

With `no-invalid-position-at-import-rule` disabled, it allows to have these lines after the mixin declarations:
https://github.com/searxng/searxng/blob/739f0355f9f0120791f6f59ed0c5499097cf20bf/searx/static/themes/simple/src/less/style-rtl.less#L57

https://github.com/searxng/searxng/blob/739f0355f9f0120791f6f59ed0c5499097cf20bf/searx/static/themes/simple/src/less/style-rtl.less#L57

I can create a commit for this change.

return42 replied:
> I can create a commit for this change.

this would be cool / thanks a lot!

dalf replied:
Done.
I've merged the other commits (I have a backup if for reason it was not a good choice).

return42 commented on searx/static/themes/simple/src/less/style.less:
```diff
@@ -33,6 +33,7 @@
   vertical-align: bottom;
   line-height: 1;
   text-decoration: inherit;
+  .ltr-transform();
```
> In LTR version:

```less
.ltr-left(@offset) {
  left: @offset;
}
```

> In this RTL version:

```less
.ltr-left(@offset) {
  right: @offset;
}
```

> and so on for the different CSS properties.

Why do we choose the prefix `.ltr-` when it stands for RTL and LTR replacements, isn't it better we use a prefix like `.text-direction-` or _shorter_ `.direction-` or anything other that fits better?  .. The prefix `.rtl-` might confuse here a litle bit.

dalf replied:
My reasoning:
* We need to describe one version: either the RTL either the LTR version. From this version, the LESS mixins can create the other version.
* `.ltr-left(@offset)` means `left: @offset` in the RTL version --> when `rtl` then `left` (and adjust when `ltr`) --> `.rtl-left`.
  The idea is make clear that `left` applies to the RTL version and it has to be changed in the LTR version.

`direction-left` is telling `left` for the RTL version implicitly or it can tell `left` for LTR version and `right` in the RTL version.

return42 reviewed (approved):
Except `/preferences` the RTL text orientation LGTM.

I think this PR is a huge improvement to what we have.

I left some comments with my thoughts ...

dalf committed:
[mod] simple theme: change stylelint configuration

* disable declaration-empty-line-before
  https://stylelint.io/user-guide/rules/list/declaration-empty-line-before/
  this change allows to mix CSS declarations and LESS mixins without empty lines:

  #something {
    display: flex;
    .ltr-left(60rem); // no mandatory empty line before this one
  }

* disable no-invalid-position-at-import-rule
  https://stylelint.io/user-guide/rules/list/no-invalid-position-at-import-rule/

  this change allows to declare some mixins and then import another .less file:
  for example:

  .ltr-left(@offset) {
    left: @offset;
  }
  @import "style.less";

dalf committed:
[build] /static

return42 reviewed (approved):
I think this PR is a huge improvement to what we have.

Lets merge this PR, the `/preferences` page can be fixed in another PR.

> I have a backup if for reason it was not a good choice

from my side is no longer need for.

mrpaulblack reviewed (approved):
LGTM :+1: 

(We should work on making LTR text in an RTL layout to be displayed as LTR in another PR)
